README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

A Flutter package provide third party login buttons and helper to use firebase auth.
# Notices

Before use this package, make sure you have already set up your Firebase and settings of below packages.
1. [google_sign_in](https://pub.dev/packages/google_sign_in)
2. [flutter_facebook_auth](https://facebook.meedu.app/docs/intro)
3. [sign_in_with_apple](https://pub.dev/packages/sign_in_with_apple)

# Features

1. Login button which already define for Google, Facebook, and Apple sign in.
2. You can pass function that executes when login is successful or failure.
3. Use LoginHelper to send user a passwordless login email.
4. Use LoginHelper to sign in or create new firebase user with email and password.

# Getting started

1. Add below lines to your project's pubspec.yaml dependencies.

```yaml
flutter_login:
    git:
      url: https://github.com/mirror-media/flutter_login.git
      ref: main
      version: ^0.0.9
```

2. Import library by adding the following line to where you want to use button or helper.
```dart
import 'package:flutter_login/flutter_login.dart';
```

# Usage 
## Login type:
Supported third-party logins
```dart
enum LoginType {
  facebook,
  google,
  apple,
}
```
## Login Status:
 Return firebase_auth login result
```dart
enum FirebaseLoginStatus {
  cancel,
  success,
  error,
}
```

## class LoginButton:

A prebuild button widget for each LoginType.

```dart
/// Define button login type. Required
final LoginType type;

/// Execute when login process finish
final LoginFinishCallback? onFinished;

/// Customize button text
/// Default text:
/// Apple: 以 Apple 帳號繼續
/// Facebook: 以 Facebook 帳號繼續
/// Google: 以 Google 帳號繼續
final String? buttonText;

/// Customize button text size. Default is 16.0
final double textSize;

/// Customize button text color. Default is black
final Color textColor;

/// Whether to show the icon. Default is true
final bool showIcon;

/// Customize button backgroundColor. Default is white
final Color buttonBackgroundColor;

/// Customize button borderColor. Default is black
final Color buttonBorderColor;

/// Customize button loading animation color. Default is black12
final Color loadingAnimationColor;

/// Whether to handle Firebase's "account-exists-with-different-credential" error
/// Default is true
/// If true, when error occur, will pop a dialog and link new credential to existing firebase user
/// To know more, please read this [page](https://firebase.google.com/docs/auth/flutter/errors)
final bool handleAccountExistsWithDifferentCredentialError;

/// Customize button icon color.
/// Default color:
/// Apple: Black
/// Facebook: RGBO(23, 120, 242, 1)
/// Google: Colorful svg
final Color? iconColor;

/// Customize button icon. Default is logo by login type
/// This will place in OutlinedButton's icon
final Widget? icon;
```

## class LoginHelper:

There are 4 main usage.

All of them are Future and will return a boolean that indicates whether or not the user is log in (or create) successfully.

<br />

### 1. Send a email with link for passwordless login:
```dart
Future<bool> signInWithEmailAndLink(String email, String link)
```
email is the email address that will be sent to, link is the user will be redirected to when there is no application.

**Notice: It will save the email as String by used SharedPreferences with key "signInEmail", you will need to get it from shared preferences when the user open the application from login email to verify.**

**Notice 2: If you want to use this sign in method, you need to open passwordless sign in in firebase console.**

<br />

### 2. Third party sign in:

Now support Google, Facebook, and Apple.

They have same optional named parameter: handleAccountExistsWithDifferentCredentialError.

- handleAccountExistsWithDifferentCredentialError: Decide whether or not handle account-exists-with-different-credential error, default is true.
<br />

signInWithGoogle:
```dart
Future<FirebaseLoginStatus> signInWithGoogle({
  bool handleAccountExistsWithDifferentCredentialError = true,
})
```
signInWithFacebook:
```dart
Future<FirebaseLoginStatus> signInWithFacebook({
  bool handleAccountExistsWithDifferentCredentialError = true,
})
```
signInWithApple:
```dart
Future<FirebaseLoginStatus> signInWithApple({
  bool handleAccountExistsWithDifferentCredentialError = true,
})
```
<br />

### 3. Email and password sign in:

There are two parameter: email and password. And optional named parameters:  ifNotExistsCreateUser, askAgain.

- ifNotExistsCreateUser: Decide if can't find user whether directly create one new user via createUserWithEmailAndPassword(), default is true.
- askAgain: Decide whether show a dialog when user password not correct, default is false. 


```dart
Future<FirebaseLoginStatus> signInWithEmailAndPassword(
  String email,
  String password, {
  bool ifNotExistsCreateUser = true,
  bool askAgain = false,
})
```

<br />

### 4. Create new user with email and password:

There are two parameter: email and password. And optional named parameters:  ifExistsTrySignIn.

- ifExistsTrySignIn: Decide if email is already exists, whether directly try sign in, default is true.

```dart
Future<FirebaseLoginStatus> createUserWithEmailAndPassword(
  String email,
  String password, {
  bool ifExistsTrySignIn = true,
})
```
<br />

### 'account-exists-with-different-credential' error

It will occur when user already exists in Firebase auth. 
<br>By default this package will handle this error by show a dialog and ask user to sign in with default sign in method provide by Firebase, then link currrent sign in method to Firebase auth.
<br> To know more, please check this [page](https://firebase.google.com/docs/auth/flutter/errors).

<br>

### Others:

It has two getter, isNewUser and signinError.

- isNewUser: It will return a boolean indicating whether user is new.

  **Notice: Only use after signIn successfully, or may get error value**


- signinError: It will return dynamic that previous sign in method catch error.

  **Notice: Only use after signIn failed, or may get null**


CHANGELOG.md
## 0.0.9

- Add one_context package to remove context in async functions
- Add optional parameter icon and iconColor to LoginButton
- Add comments to function and parameters
- Upgrade example to Flutter 3

**BREAKING CHANGE:**
- **Refactor: change paramter name from "handlingAccountExistsWithDifferentCredentialError" to "handleAccountExistsWithDifferentCredentialError"**
- **Remove context in function's parameter**

## 0.0.8

- Upgrade sign_in_with_apple to ^4.1.0
- Upgrade font_awesome_flutter to ^10.2.1
- Upgrade flutter_lints to ^2.0.1


## 0.0.6

- Upgrade flutter_facebook_auth to ^4.1.2
- Upgrade firebase_auth to ^3.3.11
- Upgrade sign_in_with_apple to ^3.3.0
- Upgrade google_sign_in to ^5.2.4
- Upgrade package_info_plus to ^1.4.0
- Upgrade flutter_svg to ^1.0.3
- Upgrade shared_preferences to ^2.0.13

**BREAKING CHANGE:**
- **Add enum FirebaseLoginStatus**
- **Change login_helper return type to enum FirebaseLoginStatus**
- **Change login_button parameter onSuccess and onFailed to onFinished**



## 0.0.5

- Upgrade flutter_facebook_auth to ^4.1.1


## 0.0.4

- Add example app connect to MirrorMedia App Dev firebase.
- Handle error when user want login that email is existed but no password.
- Ask user key password again when linking existed email.


## 0.0.3

- Parameter onFailed of LoginButton now will pass the error into function.
- LoginButton add a parameter 'buttonText' let you custom the text.
- LoginButton add a parameter 'handlingAccountExistsWithDifferentCredentialError' to decide whether handle 'account-exists-with-different-credential' error.
- Show a hint dialog when account-exists-with-different-credential error happen.
- Remove unnecessary function generateNonce in LoginHelper.
- Add createUserWithEmailAndPassword() and signInWithEmailAndPassword() in LoginHelper.
- Now can handle 'account-exists-with-different-credential' error.


## 0.0.2

- Change onSuccess in LoginButton to valueSetter with a boolean isNewUser.
- Add error handler for 'account-exists-with-different-credential' error. (BETA)


## 0.0.1

- Outlined button with three style (Google, Facebook and Apple).
- A helper that contains sign in function.


example/lib/test_app.dart
import 'dart:io';

import 'package:email_validator/email_validator.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_login/flutter_login.dart';

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'flutter_login_example',
      home: TestApp(),
    );
  }
}

class TestApp extends StatefulWidget {
  const TestApp({Key? key}) : super(key: key);
  @override
  _TestAppState createState() => _TestAppState();
}

class _TestAppState extends State<TestApp> {
  String loginStauts = "none";
  FirebaseAuth auth = FirebaseAuth.instance;
  LoginHelper loginHelper = LoginHelper();

  @override
  Widget build(BuildContext context) {
    // force portrait
    SystemChrome.setPreferredOrientations([
      DeviceOrientation.portraitUp,
      DeviceOrientation.portraitDown,
    ]);

    TextEditingController _controller1 = TextEditingController();
    TextEditingController _controller2 = TextEditingController();
    bool isLoggedIn = false;
    if (auth.currentUser != null) {
      isLoggedIn = true;
      loginStauts = auth.currentUser!.email! + '\nlog in';
    }

    return Scaffold(
      appBar: AppBar(
        title: const Text('flutter_login_example'),
      ),
      body: SafeArea(
        child: ListView(
          padding: const EdgeInsets.symmetric(horizontal: 50),
          children: [
            const SizedBox(
              height: 20,
            ),
            const Center(
              child: Text(
                'Login status',
                style: TextStyle(fontSize: 30),
              ),
            ),
            Center(
              child: Text(
                loginStauts,
                textAlign: TextAlign.center,
                style: const TextStyle(fontSize: 20),
              ),
            ),
            const SizedBox(
              height: 40,
            ),
            LoginButton(
              type: LoginType.google,
              onFinished: (status, isNewUser, error) {
                if (status == FirebaseLoginStatus.success) {
                  loginStauts = auth.currentUser!.email! + '\nlog in';
                  isLoggedIn = true;
                } else if (status == FirebaseLoginStatus.cancel) {
                  loginStauts = 'Login cancelled';
                } else {
                  loginStauts = 'Log in failed';
                }
                setState(() {});
              },
            ),
            const SizedBox(
              height: 10,
            ),
            LoginButton(
              type: LoginType.facebook,
              onFinished: (status, isNewUser, error) {
                if (status == FirebaseLoginStatus.success) {
                  loginStauts = auth.currentUser!.email! + '\nlog in';
                  isLoggedIn = true;
                } else if (status == FirebaseLoginStatus.cancel) {
                  loginStauts = 'Login cancelled';
                } else {
                  loginStauts = 'Log in failed';
                }
                setState(() {});
              },
            ),
            const SizedBox(
              height: 10,
            ),
            if (Platform.isIOS)
              LoginButton(
                type: LoginType.apple,
                onFinished: (status, isNewUser, error) {
                  if (status == FirebaseLoginStatus.success) {
                    loginStauts = auth.currentUser!.email! + '\nlog in';
                    isLoggedIn = true;
                  } else if (status == FirebaseLoginStatus.cancel) {
                    loginStauts = 'Login cancelled';
                  } else {
                    loginStauts = 'Log in failed';
                  }
                  setState(() {});
                },
              ),
            const SizedBox(
              height: 30,
            ),
            const Center(
              child: Text('Email'),
            ),
            TextField(
              controller: _controller1,
            ),
            const SizedBox(
              height: 20,
            ),
            const Center(
              child: Text('Password'),
            ),
            TextField(
              controller: _controller2,
            ),
            const SizedBox(
              height: 10,
            ),
            if (!isLoggedIn)
              ElevatedButton(
                onPressed: () async {
                  if (EmailValidator.validate(_controller1.text)) {
                    FirebaseLoginStatus status =
                        await _loginWithEmailAndPassword(
                            _controller1.text, _controller2.text);
                    if (status == FirebaseLoginStatus.success) {
                      loginStauts = auth.currentUser!.email! + '\nlog in';
                      isLoggedIn = true;
                    } else if (status == FirebaseLoginStatus.cancel) {
                      loginStauts = 'Login cancelled';
                    } else {
                      loginStauts = 'Log in failed';
                    }
                    setState(() {});
                  } else {
                    setState(() {
                      loginStauts = 'Please enter a valid email';
                    });
                  }
                },
                child: const Text('Submit'),
              ),
            if (isLoggedIn)
              ElevatedButton(
                onPressed: () async {
                  await auth.signOut();
                  setState(() {
                    isLoggedIn = false;
                    loginStauts = "none";
                  });
                },
                child: const Text('Log out'),
              )
          ],
        ),
      ),
    );
  }

  Future<FirebaseLoginStatus> _loginWithEmailAndPassword(
      String email, String password) async {
    return await loginHelper.signInWithEmailAndPassword(
      email,
      password,
      askAgain: true,
    );
  }
}


lib/src/login_button.dart
import 'package:flutter/material.dart';
import 'package:flutter_spinkit/flutter_spinkit.dart';
import 'package:flutter_svg/flutter_svg.dart';
import 'package:font_awesome_flutter/font_awesome_flutter.dart';

import 'login_helper.dart';

/// Supported third-party logins
enum LoginType {
  facebook,
  google,
  apple,
}

/// Return when login process finish with result
typedef LoginFinishCallback = void Function(
    FirebaseLoginStatus status, bool isNewUser, dynamic error);

class LoginButton extends StatefulWidget {
  /// Define button login type. Required
  final LoginType type;

  /// Execute when login process finish
  final LoginFinishCallback? onFinished;

  /// Customize button text
  /// <br>Default text:
  /// + Apple: 以 Apple 帳號繼續
  /// + Facebook: 以 Facebook 帳號繼續
  /// + Google: 以 Google 帳號繼續
  final String? buttonText;

  /// Customize button text size. Default is 16.0
  final double textSize;

  /// Customize button text color. Default is black
  final Color textColor;

  /// Whether to show the icon. Default is true
  final bool showIcon;

  /// Customize button backgroundColor. Default is white
  final Color buttonBackgroundColor;

  /// Customize button borderColor. Default is black
  final Color buttonBorderColor;

  /// Customize button loading animation color. Default is black12
  final Color loadingAnimationColor;

  /// Whether to handle Firebase's "account-exists-with-different-credential" error
  /// <br> Default is true
  /// <br> If true, when error occur, will pop a dialog and link new credential to existing firebase user
  /// <br> To know more, please read this [page](https://firebase.google.com/docs/auth/flutter/errors)
  final bool handleAccountExistsWithDifferentCredentialError;

  /// Customize button icon color.
  /// <br>Default color:
  /// + Apple: Black
  /// + Facebook: RGBO(23, 120, 242, 1)
  /// + Google: Colorful svg
  final Color? iconColor;

  /// Customize button icon. Default is logo by login type
  /// <br>This will place in OutlinedButton's icon
  final Widget? icon;

  const LoginButton({
    Key? key,
    required this.type,
    this.onFinished,
    this.buttonText,
    this.textSize = 16,
    this.textColor = Colors.black,
    this.showIcon = true,
    this.buttonBackgroundColor = Colors.white,
    this.buttonBorderColor = Colors.black,
    this.loadingAnimationColor = Colors.black12,
    this.handleAccountExistsWithDifferentCredentialError = true,
    this.iconColor,
    this.icon,
  }) : super(key: key);

  @override
  State<LoginButton> createState() => _LoginButtonState();
}

class _LoginButtonState extends State<LoginButton> {
  bool _isLoading = false;
  late String _buttonText;
  late Widget _icon;
  final LoginHelper _loginHelper = LoginHelper();

  @override
  void initState() {
    if (widget.type == LoginType.apple) {
      _buttonText = '以 Apple 帳號繼續';
      _icon = FaIcon(
        FontAwesomeIcons.apple,
        size: 18,
        color: widget.iconColor ?? Colors.black,
      );
    } else if (widget.type == LoginType.facebook) {
      _buttonText = '以 Facebook 帳號繼續';
      _icon = FaIcon(
        FontAwesomeIcons.squareFacebook,
        size: 18,
        color: widget.iconColor ?? const Color.fromRGBO(23, 120, 242, 1),
      );
    } else if (widget.type == LoginType.google) {
      _buttonText = '以 Google 帳號繼續';
      _icon = SvgPicture.asset(
        'assets/images/googleLogo.svg',
        package: 'flutter_login',
        width: 16,
        height: 16,
        color: widget.iconColor,
      );
    }

    if (!widget.showIcon) {
      _icon = Container();
    }

    if (widget.buttonText != null) {
      _buttonText = widget.buttonText!;
    }

    if (widget.icon != null) {
      _icon = widget.icon!;
    }
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return OutlinedButton.icon(
      onPressed: () async {
        setState(() {
          _isLoading = true;
        });
        FirebaseLoginStatus result;
        if (widget.type == LoginType.apple) {
          result = await _loginHelper.signInWithApple(
            handleAccountExistsWithDifferentCredentialError:
                widget.handleAccountExistsWithDifferentCredentialError,
          );
        } else if (widget.type == LoginType.facebook) {
          result = await _loginHelper.signInWithFacebook(
            handeAccountExistsWithDifferentCredentialError:
                widget.handleAccountExistsWithDifferentCredentialError,
          );
        } else {
          result = await _loginHelper.signInWithGoogle(
            handleAccountExistsWithDifferentCredentialError:
                widget.handleAccountExistsWithDifferentCredentialError,
          );
        }

        if (widget.onFinished != null) {
          widget.onFinished!(
            result,
            _loginHelper.isNewUser,
            _loginHelper.error,
          );
        }
        setState(() {
          _isLoading = false;
        });
      },
      style: OutlinedButton.styleFrom(
        backgroundColor: widget.buttonBackgroundColor,
        side: BorderSide(
          color: widget.buttonBorderColor,
          width: 1,
        ),
        fixedSize: const Size(double.infinity, 48),
      ),
      icon: _isLoading ? Container() : _icon,
      label: _isLoading
          ? SpinKitThreeBounce(color: widget.loadingAnimationColor, size: 30)
          : Text(
              _buttonText,
              style: TextStyle(
                fontSize: widget.textSize,
                fontWeight: FontWeight.w400,
                color: widget.textColor,
              ),
            ),
    );
  }
}
